x-pack/plugins/alerting/server/alerts_client/lib/build_recovered_alert.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEmpty } from 'lodash';
+import type { Alert } from '@kbn/alerts-as-data-utils';
+import { Alert as LegacyAlert } from '../../alert/alert';
+import { AlertInstanceContext, AlertInstanceState, RuleAlertData } from '../../types';
+import type { AlertRule } from '../types';
+
+interface BuildRecoveredAlertOpts<
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  alert: Alert & AlertData;
+  legacyAlert: LegacyAlert<LegacyState, LegacyContext, ActionGroupIds | RecoveryActionGroupId>;
+  rule: AlertRule;
+  timestamp: string;
+}
+
+/**
+ * Updates an existing alert document with data from the LegacyAlert class
+ * This could be a currently active alert that is now recovered or a previously
+ * recovered alert that has updates to its flapping history
+ * Currently only populates framework fields and not any rule type specific fields
+ */
+
+export const buildRecoveredAlert = <
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>({
+  alert,
+  legacyAlert,
+  rule,
+  timestamp,
+}: BuildRecoveredAlertOpts<
+  AlertData,
+  LegacyState,
+  LegacyContext,
+  ActionGroupIds,
+  RecoveryActionGroupId
+>): Alert & AlertData => {
+  return {
+    ...alert,
+    // Update the timestamp to reflect latest update time
+    '@timestamp': timestamp,
+    kibana: {
+      ...alert.kibana,
+      alert: {
+        ...alert.kibana.alert,
+        // Set latest flapping state
+        flapping: legacyAlert.getFlapping(),
+        // Set latest rule configuration
+        rule: rule.kibana?.alert.rule,
+        // Set status to 'recovered'
+        status: 'recovered',
```

Comment:
Do you have an opinion one way or the other? I was undecided on whether to update the `action_group` here since we would be losing the history of which action group the alert was active in, but I guess in the case of a rule type with multiple action groups, if an alert remained active but switched between groups, we would lose that history then anyway?

Reply:
@ymao1 and I discussed offline and we'll update the `action_group` on recovery. Which will solve the concern from https://github.com/elastic/kibana/pull/156946#discussion_r1202437740.

Reply:
Updated in https://github.com/elastic/kibana/pull/156946/commits/1f376bd2b548eab6a3905af56a9686e3042dc62b